scripts/pylib/twister/twisterlib.py

```diff
@@ -2767,6 +2770,49 @@ def process(self, pipeline, done, message, lock, results):
             else:
                 self.cleanup_artifacts()
 
+    def amend_test_instance(self):
+        lst_file = os.path.join(self.build_dir, "zephyr", "zephyr.lst")
+        if os.path.isfile(lst_file):
+            logger.debug(f"zephyr.lst located: {lst_file}")
+        else:
+            # no zephyr.lst file, cannot do symbol-based amending
+            logger.debug(f"zephyr.lst NOT located: {lst_file}")
+            return
+
+        yaml_testsuite_name = self.instance.name.split(os.path.sep)[-1]
+        logger.debug(f"amending yaml test suite: {yaml_testsuite_name}")
+        intended_testcases = []
+
+        with open(lst_file, 'r') as fp:
+            lst_content = fp.read()
+            logger.debug(f"test instance {self.instance.name} has {len(self.instance.testcases)} cases.")
+            for testcase in self.instance.testcases:
+                # identifier is "yaml_testsuite.test_function" or just "yaml_testsuite"
+                identifier = testcase.name
+                if identifier == yaml_testsuite_name:
+                    # Sometimes twsiter add yaml test suite name to a test instance as a test case.
+                    # Such test case has no real test function.
+                    # Skip it to make stats right
+                    logger.debug(f"skipping: {identifier}")
+                    continue
+
+                test_func_name1 = f"test_{identifier.split('.')[-1]}"
+                test_func_name2 = f"{identifier.split('.')[-1]}"
+                # If the .lst file exists and the test case function not in it,
+                # it should be an un-intended test function for this test instance build.
+                # If this is an old test plan and contains test cases collected at runtime
+                # by checking the log, such test cases may not have functions in .lst file.
+                # Also remove them here, which should be OK because they will still be added back at runtime.
+                if not (test_func_name1 in lst_content or test_func_name2 in lst_content):
+                    logger.debug(f"Unintended test case: {identifier}")
+                    continue
+
+                intended_testcases.append(testcase)
+                logger.debug(f"Intended test case: {identifier}")
+
+        self.instance.testcases.clear()
```

Comment:
But such combinations are totally irrational. They are included in the test plan merely because we didn't do a proper C preprocessing. If we have a chance to suppress them, why not?

Reply:
I agree with you that test plan are prepared in wrong way, due to the fact that we are not able to preprocess C files better. And generating test plan after run tests (after run building) is not the best solution - the idea of test plan is to create "plan" with information what test will be performed, to make it possible to verify at the end if all planed tests was really performed (or eventually why not). As far as I understand, the test plan should help us detect some kind of "silent" skips performed during runtime.

If you completely remove information from final report about this unintended test cases, then you get huge discrepancy between test plan and test report. And potential user can be confused - what happen with those missing testcases?

To solve this problem entirely the tests source code should be changed - to be possible to parse them properly by Twister. But I know, that this is too much to do at this moment. So, for me some kind of compromise is marking those unintended tests as "skip" - this is still better solution rather than marking them as "null".

Reply:
@PerMac Yes, the `testcases` keyword is the most reliable way so far. It bypasses the source file parsing and offloads the duty to the case author. If we can change all the existing cases as necessary and ensure coding discipline, this can be a solution. This PR is meant to mitigate the incorrect result caused by source file parsing.